File: py/953.verifying-an-alien-dictionary.py
from typing import List


class Solution:
    def isAlienSorted(self, words: List[str], order: str) -> bool:
        ch_idx = {ch: idx for idx, ch in enumerate(order)}
        for i in range(len(words)-1):
            for j in range(len(words[i])):
                if j >= len(words[i+1]):
                    return False
                if words[i][j] != words[i+1][j]:
                    if ch_idx[words[i][j]] > ch_idx[words[i+1][j]]:
                        return False
                    break
        return True

File: cpp/953.verifying-an-alien-dictionary.cpp
#include <string>
#include <unordered_map>
#include <vector>

using namespace std;

class Solution {
public:
  bool isAlienSorted(vector<string> &words, string order) {
    unordered_map<char, int> char_idx;
    for (int idx = 0; idx < order.length(); idx++) {
      char_idx[order[idx]] = idx;
    }
    for (int i = 0; i < words.size() - 1; i++) {
      for (int j = 0; j < words[i].size(); j++) {
        if (j >= words[i + 1].length()) {
          return false;
        }
        if (char_idx[words[i][j]] < char_idx[words[i + 1][j]]) {
          break;
        }
        if (char_idx[words[i][j]] > char_idx[words[i + 1][j]]) {
          return false;
        }
      }
    }
    return true;
  }
};